WordPress/src/main/java/org/wordpress/android/ui/posts/RemotePreviewLogicHelper.kt

```diff
@@ -0,0 +1,106 @@
+        REMOTE_PREVIEW_WITH_REMOTE_AUTO_SAVE
+    }
+
+    enum class PreviewLogicOperationResult {
+        PREVIEW_NOT_AVAILABLE,
+        NETWORK_NOT_AVAILABLE,
+        MEDIA_UPLOAD_IN_PROGRESS,
+        CANNOT_SAVE_EMPTY_DRAFT,
+        GENERATING_PREVIEW,
+        OPENING_PREVIEW,
+        CANNOT_REMOTE_AUTO_SAVE_EMPTY_POST
+    }
+
+    interface RemotePreviewHelperFunctions {
+        fun notifyUploadInProgress(post: PostModel): Boolean
+        fun updatePostIfNeeded(): PostModel? = null
+        fun notifyEmptyDraft() {}
+        fun startUploading(isRemoteAutoSave: Boolean, post: PostModel)
+        fun notifyEmptyPost() {}
+    }
+
+    fun runPostPreviewLogic(
+        activity: Activity,
+        site: SiteModel,
+        post: PostModel,
+        helperFunctions: RemotePreviewHelperFunctions
+    ): PreviewLogicOperationResult {
+        if (!site.isUsingWpComRestApi) {
+            activityLauncherWrapper.showActionableEmptyView(
+                    activity,
+                    WPWebViewUsageCategory.REMOTE_PREVIEW_NOT_AVAILABLE,
+                    post.title
+            )
+            return PreviewLogicOperationResult.PREVIEW_NOT_AVAILABLE
+        } else if (!networkUtilsWrapper.isNetworkAvailable()) {
+            activityLauncherWrapper.showActionableEmptyView(
+                    activity,
+                    WPWebViewUsageCategory.REMOTE_PREVIEW_NO_NETWORK,
+                    post.title
+            )
+            return PreviewLogicOperationResult.NETWORK_NOT_AVAILABLE
+        } else {
+            if (helperFunctions.notifyUploadInProgress(post)) {
+                return PreviewLogicOperationResult.MEDIA_UPLOAD_IN_PROGRESS
+            }
+
+            val updatedPost = helperFunctions.updatePostIfNeeded() ?: post
+
+            if (shouldSave(updatedPost)) {
+                if (!postUtilsWrapper.isPublishable(updatedPost)) {
+                    helperFunctions.notifyEmptyDraft()
+                    return PreviewLogicOperationResult.CANNOT_SAVE_EMPTY_DRAFT
+                }
+
+                helperFunctions.startUploading(false, updatedPost)
+            } else if (shouldRemoteAutoSave(updatedPost)) {
+                if (!postUtilsWrapper.isPublishable(updatedPost)) {
+                    helperFunctions.notifyEmptyPost()
+                    return PreviewLogicOperationResult.CANNOT_REMOTE_AUTO_SAVE_EMPTY_POST
+                }
+                helperFunctions.startUploading(true, updatedPost)
+            } else {
+                activityLauncherWrapper.previewPostOrPageForResult(
+                        activity,
+                        site,
+                        updatedPost,
+                        RemotePreviewType.REMOTE_PREVIEW
+                )
+                return PreviewLogicOperationResult.OPENING_PREVIEW
+            }
+        }
+        return PreviewLogicOperationResult.GENERATING_PREVIEW
+    }
+
+    private fun shouldSave(post: PostModel): Boolean {
```

Comment:
Sorry I haven't done that comment in my first review pass, but can we rename `shoulSave()` method to `shouldUpload()` just to make sure we don't mix it up with "local" save.

Reply:
Yep, you are right, it makes sense. Changed here 4839b9a04e50c0b4f9683ea634fa7712e687959f